### PR TITLE
[RPC] Improve error/promotion handling

### PR DESCRIPTION
**Pull request by rgrinberg**
* Add severity to error notifications. Should these be renamed to
diagnostics perhaps?

### CODE REVIEW

**Comment by rgrinberg** [comment]
@tmattio could you look over the interface and see if we're missing something?

**Comment by rgrinberg** [comment]
I renamed it to `Diagnostic`. 

I'm also planning to extend the API to cover:

* Start/End of build events
* Progress reporting for the build

File: otherlibs/dune-rpc/private/dune_rpc_private.mli [hunk 1]
```diff
@@ -121,11 +121,35 @@ module Loc : sig
...
+
   type t =
     { target : string option
     ; message : string
     ; loc : Loc.t option
+    ; severity : severity option
+    }
+end
+
+module Build : sig
```
Comment (rgrinberg) [inline otherlibs/dune-rpc/private/dune_rpc_private.mli:137]: @tmattio have a look and see if this signature covers everything we need. I'll then follow up with an implementation of these rpc calls.

Comment (tmattio) [reply otherlibs/dune-rpc/private/dune_rpc_private.mli:137]: I don't have a good idea of the requirements at the moment, so I would suggest that we start by listing the use cases in `vscode`/`lsp` that we want to implement using Dune RPC, and derive the spec of requests from them. I can start working on that next week.

Without these, one requirement that I see is to have the list of output files/artifacts for the build. This could be useful for the work on PPX analysis and will help in implementing build/debug commands. Not sure if this signature is the right place for it though. Maybe this could also be used to fix https://github.com/ocamllabs/vscode-ocaml-platform/issues/524

Comment (rgrinberg) [reply otherlibs/dune-rpc/private/dune_rpc_private.mli:137]: For now, I think it's enough to focus on the watch mode integration. The other features are much less important.

> Without these, one requirement that I see is to have the list of output files/artifacts for the build. This could be useful for the work on PPX analysis and will help in implementing build/debug commands. Not sure if this signature is the right place for it though.

It's the right place to do it.  For ppx, I suppose we could add a declaration like:

```ocaml
val pped : (string, string) request (* module path -> preprocessed output *)
```

To give the preprocessed output. It's something that will need to be delayed until watch mode is done though.

> Maybe this could also be used to fix ocamllabs/vscode-ocaml-platform#524

Let's see what @voodoos thinks about this issue first

**Comment by rgrinberg** [comment]
@cwong-ocaml I'm adding you as a reviewer since you'll be using this API as well. In particular, take a look at the changes to`dune_rpc.mli` to make sure that the API makes sense.

File: otherlibs/dune-rpc/dune_rpc.mli [hunk 2]
```diff
@@ -82,25 +82,56 @@ module V1 : sig
...
+        ; in_source : string
+        }
+    end
+
     type t =
       { target : string option
       ; message : string
       ; loc : Loc.t option
+      ; severity : severity option
+      ; promotion : Promotion.t option
```
Comment (tmattio) [inline otherlibs/dune-rpc/dune_rpc.mli:102]: Why are promotions in diagnostic now? It seems unrelated, at least from a user's point of view.

File: otherlibs/dune-rpc/private/dune_rpc_private.mli [hunk 3]
```diff
@@ -121,32 +121,64 @@ module Loc : sig
...
+      }
+  end
+
   type t =
     { target : string option
     ; message : string
     ; loc : Loc.t option
+    ; severity : severity option
+    ; promotion : Promotion.t option
+          (* TODO should this be generalied to a list? *)
```
Comment (tmattio) [inline otherlibs/dune-rpc/private/dune_rpc_private.mli:142]: What is the alternative? Do you send one diagnostic per promotion?

**Review by cwong-ocaml — COMMENTED** [review]
I don't have much to add beyond the comments raised by @tmattio about the API. Having promotions as a diagnostic seems a bit strange to me as well; intuitively I'd have thought of it more as an error-like event.

File: otherlibs/dune-rpc/private/dune_rpc_private.mli (hunk 3 above)
Comment (rgrinberg) [reply otherlibs/dune-rpc/private/dune_rpc_private.mli:142]: Some users asked to have more than 1 promotion per failed rule. E.g. `(progn (diff ..) (diff ..))`, but without failing on the first `diff`.

File: otherlibs/dune-rpc/dune_rpc.mli (hunk 2 above)
Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.mli:102]: I think we came up with this idea in a meeting with @jeremiedimino. The idea is that a promotion should be counted as an error because it suggests that some `diff` action failed the build.

**Comment by rgrinberg** [comment]
> Having promotions as a diagnostic seems a bit strange to me as well; intuitively I'd have thought of it more as an error-like event.

A diagnostic is an "error-like" event. Though it's not necessarily an error as it could also be a warning.

**Comment by rgrinberg** [comment]
However `diagnostic.severity = Warning` and `diagnostic.promotion = Some _` are mutually exclusive (at least at the moment), and the current design reflect that. Let's have @jeremiedimino take a quick glance at the mli

File: otherlibs/dune-rpc/dune_rpc.mli [hunk 4]
```diff
@@ -82,25 +82,56 @@ module V1 : sig
...
+    module Event : sig
+      type nonrec t =
+        | Add of t
+        | Remove of t
+    end
   end
 
-  module Promotion : sig
+  module Build : sig
+    type t
```
Comment (cwong-ocaml) [inline otherlibs/dune-rpc/dune_rpc.mli:113]: Is the intention to eventually add functions consuming a `Build.t` to this module? At the moment, the optional `build_event` and `build_progress` arguments to `create` can't actually do anything other than ignore that argument.

Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.mli:113]: Yes, I'm thinking of adding functions such as:

```ocaml
val targets : Build.t -> target list
```

**Comment by jeremiedimino** [comment]
If `{severity = Warning; promotion = Some _; _}` is an impossible case, the data types should reflect it. It is generally good to make data types tights, so that impossible case cannot be represented and we get absolute invariant about data structures.

**Comment by jeremiedimino** [comment]
Just had a chat with Rudi. IIUC, "diagnostic" is the terminology used by VSCode, which is the motivation for having errors and warnings are under the same umbrella.

Having the editor knows about promotions is nice has the editor might have better diffing tools, so we want Dune to pass this information to the editor.

That leaves us with the case with promotions attached to a warning. In the end, while Dune doesn't produce such cases at the moment, it doesn't seem like an absurd case either. We could imagine a command suggesting a correction but not failing. So the current design seems fine to me.

**Comment by rgrinberg** [comment]
I just realized that I forgot to update the API to cover the issue of "flashing errors". This occurs when when a new build starts, the old errors are cleared, and the new errors are emitted that are usually the same as the old ones. This is quite annoying user experience wise, an the api should be careful to invalidate existing errors more carefully.

**Comment by rgrinberg** [comment]
@jeremiedimino I brought back manually clearing errors. Since the build system now has a callback to remove errors, I think it makes sense to remove the errors after the notification.

File: otherlibs/dune-rpc/dune_rpc.mli [hunk 5]
```diff
@@ -76,25 +76,60 @@ module V1 : sig
...
+    val severity : t -> severity option
+
+    val promotion : t -> Promotion.t list
+
+    val targets : t -> string list
+
+    module Event : sig
+      type nonrec t =
+        | Add of t
+        | Remove of t
```
Comment (cwong-ocaml) [inline otherlibs/dune-rpc/dune_rpc.mli:106]: It might be better to associate a diagnostic with a unique (within a session, at least) id and make the message `Remove of Id.t`.

Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.mli:106]: Yup, I am thinking of something like that.

> within a session, at least

This is what I'm wondering about. I'm thinking just making the ID the hash of all the fields. I'd prefer if the Id's persist across sessions as otherwise it's a headache for clients to always make sure they don't store stale Id's.

>  id and make the message Remove of Id.t.

What about just making this transparent? The user passes the error, but the protocol passes the ID under the hood. Id's sometimes have tricky semantics, and I hate letting the user know of them.

Comment (cwong-ocaml) [reply otherlibs/dune-rpc/dune_rpc.mli:106]: >  I'd prefer if the Id's persist across sessions as otherwise it's a headache for clients to always make sure they don't store stale Id's.

This makes a lot of sense :+1:

> The user passes the error, but the protocol passes the ID under the hood.

I'm of two minds about this. Normally, I am certainly all for keeping this kind of thing transparent for exactly the reason you suggest. On the other hand, without the id, if a client is maintaining a set of diagnostics, handling a `Remove t` event would require inspecting all of the error fields.

**Comment by rgrinberg** [comment]
@jeremiedimino it seems like we need something like the fiber_detached construct we discussed long ago (in the context of using fibers in LSP). This is the interface:

```ocaml
type t

val create : unit -> t

val task : t -> f:(unit -> unit Fiber.t) -> unit

val run : t -> unit Fiber.t
```

Why do we need it? When build system encounters an error, it calls the appropriate `Error.t -> unit` callback in `Build_system.Handler.t`. In turn, the RPC server will then send  this error to subscribers in a fiber valued function. But then, where do we wait for this fiber?

**Comment by jeremiedimino** [comment]
The only problem I see with this API is `task` returning `unit` rather than `unit Fiber.t`. It does a side-effect so it should return `Fiber.t`. If you make `task` return `Fiber.t`, then it's going to be difficult to connect it to `with_error_handler`. Seems like we need an alternative version of `with_error_handler` than takes a callback returning a fiber.

**Comment by rgrinberg** [comment]
> The only problem I see with this API is task returning unit rather than unit Fiber.t. It does a side-effect so it should return Fiber.t

Note that it would still be ignored.

>  If you make task return Fiber.t, then it's going to be difficult to connect it to with_error_handler. Seems like we need an alternative version of with_error_handler than takes a callback returning a fiber.

Yeah, it would be good to discuss error handling thoroughly tomorrow.

File: otherlibs/dune-rpc/dune_rpc.mli [hunk 6]
```diff
@@ -173,8 +210,9 @@ module V1 : sig
 
       val create :
            ?log:(Message.t -> unit Fiber.t)
-        -> ?errors:(Error.t list -> unit Fiber.t)
-        -> ?promotions:(Promotion.t list -> unit Fiber.t)
+        -> ?diagnostic:(Diagnostic.Event.t -> unit Fiber.t)
```
Comment (cwong-ocaml) [inline otherlibs/dune-rpc/dune_rpc.mli:213]: If a client connects to a build midway through, will they only receive errors from that point onwards? Presumably, they can get the initial diagnostic state via the `diagnostics` request, but it might be friendly to also give this information to the handler.

Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.mli:213]: > Presumably, they can get the initial diagnostic state via the diagnostics request

Yeah, that is how I expected it to work.

>  but it might be friendly to also give this information to the handler.

How would this information be given?

Comment (cwong-ocaml) [reply otherlibs/dune-rpc/dune_rpc.mli:213]: One way might be to adjust handlers to include `initial_diagnostics_state:(Diagnostic.t list -> unit Fiber.t)`, and then the server sends that information when the session is first invoked.

That said, this isn't a crucial feature by any means.

Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.mli:213]: I'd rather not have two ways of doing the same thing unless there's clear trade-offs between the 2 mechanisms.

Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.mli:213]: The race is quite problematic. Then I suppose I'll change the subscription mechanism as you advised, and get rid of the request. I don't want to add requests unless they are strictly useful to implement a feature.

Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.mli]: > One way might be to adjust handlers to include initial_diagnostics_state:(Diagnostic.t list -> unit Fiber.t), and then the server sends that information when the session is first invoked.

Actually, I'm wondering if we should just reuse the exiting callback:
```ocaml
...
        -> ?diagnostic:(Diagnostic.Event.t list -> unit Fiber.t)
...
```

And we'll guarantee that initially we'll just initially send the initial subscriptions as `Add err` events.

Comment (cwong-ocaml) [reply otherlibs/dune-rpc/dune_rpc.mli]: Sure, that seems like a reasonable idea.

**Comment by rgrinberg** [comment]
Me and @jeremiedimino discussed this PR a little. I'll discuss some conclusions we made and some new problems I encountered.

* Dune is currently doing deduplication when displaying errors. We will not be doing this for RPC.
* To serialize error messages, we will encode `Pp.t` values directly. Client will have full freedom on how to display these values.
* Dune will send two types of errors to the client. Errors where the "User" is responsible, and errors where the "Developer" is responsible. The latter will include back traces.

Onto some problems:

* We need to make `'a Pp.t` public. I will do so in `Pp.Ast.t`. We will also need to make `Dune_rpc` depend on `Pp` (not a problem).
* The full formatting of errors requires: `Style.t Pp.t`:
```ocaml
module Style : sig
  type t =
    | Loc
    | Error
    | Warning
    | Kwd
    | Id
    | Prompt
    | Details
    | Ok
    | Debug
    | Success
    | Ansi_styles of Ansi_color.Style.t list
end
```

We can move this type `Dune_rpc`, but where should `Ansi_color` go? Shall we move it into `Dune_rpc`, or make it its own library?

**Comment by cwong-ocaml** [comment]
I'm alright with moving `Ansi_color` into its own library (maybe a more general-purpose `stdune-public` with those kinds of generally useful things -- we can also put a `Path`-like module there, to make things like locations less stringly-typed). Part of me would rather make `Ansi_color.Style.t` into some kind of sum type (or something generally more structured) but maybe it isn't worth the effort.

**Comment by rgrinberg** [comment]
Yes, those are both good points. It seems like a rich path type would be helpful to users of RPC. The distinction between Source, Build, and other paths should be useful. 

As for colors, for now let's table this discussion. This PR should be usable already without them. The build progress stuff isn't working yet, but I don't think that's blocking you.

**Comment by cwong-ocaml** [comment]
> As for colors, for now let's table this discussion. This PR should be usable already without them. The build progress stuff isn't working yet, but I don't think that's blocking you.

Indeed, this can be punted. I wouldn't go so far as to say that we're blocked by the lack of progress reporting, but is there anything I can do to help? I think we'd be pretty sad to ask people within Jane Street to test dune with a dummy progress report (currently, Ecaml reports a "0/1 targets completed, unknown status" type message), but if it's proving particularly troublesome it won't be the end of the world.

**Comment by rgrinberg** [comment]
>  I think we'd be pretty sad to ask people within Jane Street to test dune with a dummy progress report (currently, Ecaml reports a "0/1 targets completed, unknown status" type message), but if it's proving particularly troublesome it won't be the end of the world.

Actually, I can make it work quickly. The main issue is that is that I think it will require debouncing, but we can see how it works without debouncing first.

**Comment by rgrinberg** [comment]
@cwong-ocaml Alright, I've added progress notifications. For now they aren't debounced, so I'm curious how well they'll perform in Emacs.

**Comment by rgrinberg** [comment]
@jeremiedimino this is ready for review. Your review is needed mostly because of the changes to build_system, @cwong-ocaml can review the rest.

File: otherlibs/dune-rpc/dune_rpc.mli [hunk 7]
```diff
@@ -135,8 +170,9 @@ module V1 : sig
 
       val create :
            ?log:(Message.t -> unit fiber)
-        -> ?errors:(Error.t list -> unit fiber)
-        -> ?promotions:(Promotion.t list -> unit fiber)
+        -> ?diagnostic:(Diagnostic.Event.t list -> unit fiber)
```
Comment (cwong-ocaml) [inline otherlibs/dune-rpc/dune_rpc.mli:173]: Should we add some documentation guaranteeing that the first call to this callback will include every currently-existing error being `Add`ed? If we're also removing the `diagnostics` request, it might be a bit confusing for clients who would like to connect once, get the list of errors, then immediately disconnect.

**Review by cwong-ocaml — CHANGES_REQUESTED** [review]

**Comment by rgrinberg** [comment]
@jeremiedimino ping on reviewing the build_system changes :)

File: src/dune_engine/build_system.mli [hunk 8]
```diff
@@ -12,6 +12,32 @@ type caching =
   ; check_probability : float
   }
 
+module Error : sig
+  type t
+
+  val message : t -> User_message.t
+end
+
+module Handler : sig
```
Comment (jeremiedimino) [inline src/dune_engine/build_system.mli:21]: Documentation please. One should be able to understand what `Start` or `Fail` mean without having to look at the implementation.

File: otherlibs/dune-rpc/dune_rpc.mli (hunk 7 above)
Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.mli:173]: > Should we add some documentation guaranteeing that the first call to this callback will include every currently-existing error being Added?

Will do.

> If we're also removing the diagnostics request, it might be a bit confusing for clients who would like to connect once, get the list of errors, then immediately disconnect.

If there's demand for clients to do this, I will add it back.

File: src/dune_rpc_impl/server.ml [hunk 9]
```diff
@@ -60,15 +69,48 @@ end)
...
+
+  let add t which session =
+    modify t which ~f:(fun set -> Session_set.add set session)
+
+  let remove t which session =
+    modify t which ~f:(fun set -> Session_set.remove set session)
+
+  let notify t which ~f =
+    let set = get t which in
+    Fiber.parallel_iter_set (module Session_set) set ~f
```
Comment (jeremiedimino) [inline src/dune_rpc_impl/server.ml:99]: Shouldn't all this stuff be the job of the RPC? It seems odd to me that the build system has to care about the list of client that are connected.

**Review by jeremiedimino — CHANGES_REQUESTED** [review]

File: src/dune_rpc_impl/server.ml (hunk 9 above)
Comment (rgrinberg) [reply src/dune_rpc_impl/server.ml:99]: I guess I don't consider `dune_rpc_impl/server` to be part of the build system. I could move this higher up to `dune_rpc_server`.

Comment (jeremiedimino) [reply src/dune_rpc_impl/server.ml:99]: Well, here we have a property that multiple clients can subscribe to, in order to be notified when it changes. The property in question here is the of current build errors.

Subscribing to a property is a general concept. If we want to declare another property somewhere else in Dune, are we going to copy and paste this code?

It seems to me that the notion of subscribing to a property is general enough that it should be nicely wrapped up in the RPC library, so that in `build_system.ml` all we do is set the new value and be done with it. There is an additional twist here: the value is big so we send delta rather than complete values, but that seems like a general enough concept as well to me.

For instance, in D-Bus there is also a notion of property that can change over time. In the code that implement a property, all you do is declare the property and then call `set` on it. Who should be notified of this `set` is the buisiness of the D-Bus library. I would expect this to be the same with other IPC systems as well.

**Comment by jeremiedimino** [comment]
I find the callback based API in `Build_system` difficult to reason about. For instance, what should `Build_system` do if the `build_progress` callback raise an exception? It doesn't feel right to crash the build system because of a RPC failure, and I have no idea what we would do with the exception if we were catching it at this level.

Could we instead have an API where the build system sends asynchronous messages? For instance:

```
val Build_system.events : unit -> Event.t Fiber.Stream.In.t
```

and whenever we want to send an event from `build_system.ml`, we'd simply write it to this stream and be done with it. No need to worry about failure cases, these would have to be handled by the stream reader.

File: src/dune_rpc_impl/server.ml (hunk 9 above)
Comment (rgrinberg) [reply src/dune_rpc_impl/server.ml:99]: > It seems to me that the notion of subscribing to a property is general enough that it should be nicely wrapped up in the RPC library, so that in build_system.ml all we do is set the new value and be done with it. There is an additional twist here: the value is big so we send delta rather than complete values, but that seems like a general enough concept as well to me.

Adding new types of subscriptions is easy enough. You just add a new field to Subscribers.t and modify `get`.

> For instance, in D-Bus there is also a notion of property that can change over time. In the code that implement a property, all you do is declare the property and then call set on it. Who should be notified of this set is the buisiness of the D-Bus library. I would expect this to be the same with other IPC systems as well.

Yes, that's a good point and it would be much nicer to have signal semantics (signal in the sense of `React.S`). However, I think that's a pretty big change and we should first discuss the API in an issue.

**Comment by jeremiedimino** [comment]
That's a good idea. Why do we care about the push back BTW? And if there is no one reading, i would expect that the build system just write and the events get dropped.

**Comment by jeremiedimino** [comment]
To be clear: I would prefer no push back. It's also not clear what the push back means here:
- it could mean the other process received the message, but we definitely don't want to block the build system for that long
- it could mean that the RPC system received the event, but I don't see why we would care about this in the build system

**Comment by rgrinberg** [comment]
Yes, that makes sense. Let me try implementing it with `Fiber.Stream.Out.t` then.

**Comment by rgrinberg** [comment]
I cherry picked all the non controversial bits about message formatting and diagnostic types (the changes to Dune_rpc itself) that were already approved. I have a PR in the works that needs them, and I'd like to avoid more rebasing. This PR now just includes the changes to Build_system that are being discussed.

**Comment by rgrinberg** [comment]
Summarizing my discussion with Jeremie about this:

We need a dedicated abstraction to notify clients about the values changing with time. The API we're looking for is:

```ocaml
module Observable : sig
  type 'a t

  val create : unit -> 'a t * ('a -> unit Fiber.t)

  module Sub : sig
    type t
    val unsubscribe : t -> unit
  end

  val subscribe : 'a t -> ('a -> unit Fiber.t) -> Sub.t
```

It's a little like what we already have in memo, but push based. In memo, one must keep calling a memoized function to get the up to date version, here this is managed by a subscription.

**Comment by cwong-ocaml** [comment]
@rgrinberg @jeremiedimino is the plan to make these changes in this PR? It's going to be tricky testing our editor integration without at least the changes to `Build_system` that actually pass errors to the RPC.

**Review by jeremiedimino — APPROVED** [review]
Looks good. In the long run, it would be better not to inject general callbacks in the core of the build systen, but that will do for now

File: src/dune_engine/build_system.ml [hunk 10]
```diff
@@ -2185,14 +2220,28 @@ let process_exn_and_reraise exn =
...
   let f () =
-    Memo.Build.run
-      (Memo.Build.with_error_handler f ~on_error:process_exn_and_reraise)
+    let* () = t.handler.build_event Start in
+    let open Fiber.O in
+    let* res =
+      Memo.Build.run
+        (Memo.Build.with_error_handler f ~on_error:process_exn_and_reraise)
+    in
+    let+ () = t.handler.build_event Finish in
```
Comment (jeremiedimino) [inline src/dune_engine/build_system.ml:2243]: We never send this event if the build fails. Is that expected?

**Comment by jeremiedimino** [comment]
I rebased and re-worked the code in build_system.ml so that all the calls to a handler are in the fiber monad where side-effects are allowed.

File: src/dune_engine/build_system.ml (hunk 10 above)
Comment (rgrinberg) [reply src/dune_engine/build_system.ml:2190]: Yes, that is expected. There supposed to be a `Fail` event in such a case. I will add it
